src/lib/optimizer/strategy/predicate_merge_rule.cpp

```diff
@@ -0,0 +1,127 @@
+#include "logical_query_plan/union_node.hpp"
+
+using namespace opossum::expression_functional;  // NOLINT
+
+namespace opossum {
+
+void PredicateMergeRule::apply_to(const std::shared_ptr<AbstractLQPNode>& root) const {
+  Assert(root->type == LQPNodeType::Root, "PredicateMergeRule needs root to hold onto");
+
+  // For every node, store a pointer to the root of the subplan if this subplan can be merged
+  std::map<const std::shared_ptr<AbstractLQPNode>, const std::shared_ptr<AbstractLQPNode>> subplan_roots;
+
+  // For every subplan root, store the number of UnionNodes in the subplan
+  std::map<const std::shared_ptr<AbstractLQPNode>, size_t> union_node_counts;
+
+  visit_lqp(root, [&](const auto& node) {
+    if (node->type == LQPNodeType::Predicate || node->type == LQPNodeType::Union) {
+      // Add node to subplans
+      const auto& outputs = node->outputs();
+      const auto parent =
+          std::find_if(outputs.begin(), outputs.end(), [&](const auto& output) { return subplan_roots.count(output); });
+      if (parent == outputs.end()) {
+        // New subplan root found
+        subplan_roots.insert(std::make_pair(node, node));
+        union_node_counts.insert(std::make_pair(node, 0));
+      } else {
+        // New node for a known subplan found
+        subplan_roots.insert(std::make_pair(node, subplan_roots[*parent]));
+      }
+      if (node->type == LQPNodeType::Union) {
+        union_node_counts[subplan_roots[node]]++;
+      }
+    }
+    return LQPVisitation::VisitInputs;
+  });
+
+  for (const auto& node_and_count : union_node_counts) {
+    if (node_and_count.second >= optimization_threshold) {
+      _merge_subplan(node_and_count.first, std::nullopt);
+    }
+  }
+}
+
+/**
+ * Merge a subplan that only consists of PredicateNodes and UnionNodes into a single PredicateNode. The
+ * subsequent_expression parameter passes the translated expressions to the translation of its children nodes, which
+ * allows to add the translated expression of child node before its parent node to the output expression.
+ *
+ * A subplan consists of linear "chain" and forked "diamond" parts.
+ *
+ * EXAMPLE:
+ *         Step 1                   Step 2                   Step 3                         Step 4
+ *
+ *           |                        |                        |                              |
+ *      ___Union___              ___Union___           Predicate (A OR B)      Predicate ((D AND C) AND (A OR B))
+ *    /            \            /           \                  |                              |
+ * Predicate (A)   |         Predicate (A)  |                  |
+ *    |            |           |            |                  |
+ *    |       Predicate (B)    |      Predicate (B)            |
+ *    \           /            \          /                    |
+ *     Predicate (C)          Predicate (D AND C)     Predicate (D AND C)
+ *           |                        |                        |
+ *     Predicate (D)
```

Comment:
In general, it is even necessary because conjunctive predicates might be part of a disjunctive predicate (e.g., `(A AND B) OR C`). In this case, we can only merge the disjunction if we merged the conjunction before. However, I have not evaluated the case that a predicate chain is not inside a diamond, such as in the drawn example.